keps/sig-api-machinery/3396-websocket-protocol-enhancements/README.md

```diff
@@ -0,0 +1,869 @@
+protocol that API server implements on top of WebSocket has
+[deficiencies](https://github.com/kubernetes/kubernetes/issues/89899) that block the transition.
+This KEP aims to resolve them to unblock the migration and to allow all API consumers to use WebSocket rather than SPDY.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+API server defines an application-level protocol on top of WebSocket connection. It's a framing protocol
+that allows to multiplex several logical bidirectional channels on top of a single WebSocket connection.
+
+Deficiencies in the current (v4) WebSocket protocol:
+
+- No mechanism to do a half-close of a logical channel. Neither client nor server can indicate that no
+  more data is coming on a particular channel. This makes it impossible to implement certain use-cases in those
+  kubectl commands. For example, `kubectl cp file pod` spawns `tar` in the target pod and streams the contents of a file
+  as a tar stream via STDIN to the `tar` command. It needs to close STDIN to indicate that all the data has been sent.
+  If it cannot do it, `tar` gets stuck forever, waiting for more data.
+
+- No mechanism to do a reset of a logical channel to stop the other party from sending more data. This is needed for
+  parity with SPDY and is used mostly in error handling situations.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Define a new WebSocket protocol version (v5) without the deficiencies described above.
+- Change kubectl to use new WebSocket protocol for `exec`, `cp`, and `attach`. It should fall back to SPDY when
+  new protocol is not supported by the server.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- Removal of SPDY in any part of the system. In particular, API server will still have to support SPDY for some time
+  for backwards compatibility with older clients.
+- Moving `port-forward` to WebSocket. It's not clear how `port-forward` via WebSocket should work. This can be done
+  separately later.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
```

Comment:
```suggestion

- Update the WebSocket support in the API server. Support a v5 framing protocol on top of WebSocket.
```
?

Reply:
My understanding is that this section describes what changes this KEP is proposing. What you are suggesting is HOW to implement those changes, so I think it doesn't need to be here 🤷